framework/src/main/java/org/checkerframework/framework/util/FlowExpressionParseUtil.java

```diff
@@ -119,935 +102,452 @@ private static Pattern anchored(String regex) {
-
-        return Pair.of(receiver, remaining);
+        return updatedExpression;
     }
 
-    private static Receiver parseMemberSelect(
-            String s, ProcessingEnvironment env, FlowExpressionContext context, TreePath path)
-            throws FlowExpressionParseException {
-        Pair<String, String> select = parseMemberSelect(s);
-        assert select != null : "isMemberSelect must be called first";
-
-        Receiver receiver;
-        String memberSelected;
-
-        Resolver resolver = new Resolver(env);
-
-        // Attempt to match a package and class name first.
-        Pair<ClassName, String> classAndRemainingString =
-                matchPackageAndClassNameWithinExpression(s, resolver, path);
-        if (classAndRemainingString != null) {
-            receiver = classAndRemainingString.first;
-            memberSelected = classAndRemainingString.second;
-            if (memberSelected == null) {
-                throw constructParserException(
-                        s, "a class cannot terminate a flow expression string");
-            }
-        } else {
-            String receiverString = select.first;
-            memberSelected = select.second;
-            receiver = parseHelper(receiverString, context, path);
-        }
+    private static class ExpressionToReceiverVisitor
+            extends GenericVisitorWithDefaults<Receiver, FlowExpressionContext> {
 
-        if (memberSelected.equals("class")) {
-            if (receiver instanceof FlowExpressions.ClassName && !context.parsingMember) {
-                return receiver;
-            } else {
-                throw constructParserException(s, "class is not a legal identifier");
-            }
-        }
+        private final TreePath path;
+        private final ProcessingEnvironment env;
+        private final Types types;
 
-        // Parse the rest, with a new receiver.
-        FlowExpressionContext newContext = context.copyChangeToParsingMemberOfReceiver(receiver);
-        return parseHelper(memberSelected, newContext, path);
-    }
-
-    // ########
-
-    private static boolean isNullLiteral(String s, FlowExpressionContext context) {
-        if (context.parsingMember) {
-            return false;
+        ExpressionToReceiverVisitor(TreePath path, ProcessingEnvironment env) {
+            this.path = path;
+            this.env = env;
+            this.types = env.getTypeUtils();
         }
-        return s.equals("null");
-    }
 
-    private static Receiver parseNullLiteral(Types types) {
-        return new ValueLiteral(types.getNullType(), (Object) null);
-    }
-
-    private static boolean isIntLiteral(String s, FlowExpressionContext context) {
-        if (context.parsingMember) {
-            return false;
+        @Override
+        public Receiver defaultAction(
+                com.github.javaparser.ast.Node n, FlowExpressionContext context) {
+            String message = "is an unrecognized expression";
```

Comment:
I think `is not a supported expression` is better than `is an unrecognized expression`.  The expression parsed correctly; it just isn't supported.